client-s3/src/main/kotlin/app/cash/backfila/client/s3/internal/S3DatasourceBackfillOperator.kt

```diff
@@ -0,0 +1,184 @@
+package app.cash.backfila.client.s3.internal
+
+import app.cash.backfila.client.s3.S3DatasourceBackfill
+import app.cash.backfila.client.s3.shim.S3Service
+import app.cash.backfila.client.spi.BackfilaParametersOperator
+import app.cash.backfila.client.spi.BackfillOperator
+import app.cash.backfila.protos.clientservice.GetNextBatchRangeRequest
+import app.cash.backfila.protos.clientservice.GetNextBatchRangeResponse
+import app.cash.backfila.protos.clientservice.GetNextBatchRangeResponse.Batch
+import app.cash.backfila.protos.clientservice.KeyRange
+import app.cash.backfila.protos.clientservice.PrepareBackfillRequest
+import app.cash.backfila.protos.clientservice.PrepareBackfillResponse
+import app.cash.backfila.protos.clientservice.RunBatchRequest
+import app.cash.backfila.protos.clientservice.RunBatchResponse
+import com.google.common.base.Stopwatch
+import java.util.concurrent.TimeUnit
+import okio.ByteString.Companion.encodeUtf8
+
+class S3DatasourceBackfillOperator<R : Any, P : Any>(
+  override val backfill: S3DatasourceBackfill<R, P>,
+  private val parametersOperator: BackfilaParametersOperator<P>,
+  private val s3Service: S3Service,
+) : BackfillOperator {
+
+  override fun name(): String = backfill.javaClass.toString()
+
+  override fun prepareBackfill(request: PrepareBackfillRequest): PrepareBackfillResponse {
+    val config = parametersOperator.constructBackfillConfig(request)
+    backfill.validate(config)
+
+    require(request.range?.start == null && request.range?.end == null) {
+      // We could think about supporting this later by making the range mean a byte seek into all S3 files.
```

Comment:
doesn't make much sense to me to have a byte range into s3 files

Reply:
If you completed part of the file and just want to continue? I'm open to thinking about it but it is pretty low as a priority.